File: 224.basic-calculator.java
/*
 * @lc app=leetcode id=224 lang=java
 *
 * [224] Basic Calculator
 */

// @lc code=start
// Reference: https://www.cnblogs.com/grandyang/p/4570699.html
class Solution {
    // Solution 1: Iteration + stack
    public int calculate(String s) {
        return calculate1(s);
    }

    // Solution 1: Iteration + stack
    private int calculate1(String s) {
        Stack<Integer> stack = new Stack<Integer>();

        int sign = 1;   // 1: positive; 0: negative
        int num = 0;    // for digits
        int res = 0;    // for reuslt
        System.out.printf("String size: %d", stack.size());
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c >= '0') {
                num = (num * 10) + (c - '0');
                System.out.printf("num: %d", num);
            } else if (c == '+' || c == '-') {
                // after assigning the result of num * sign to res, needs let num to be 0
                res += num * sign;
                num = 0;
                sign = (c == '+') ? 1 : -1;
            } else if (c == '(') {
                // set res and sign to be default value after we push their value to stack
                stack.push(res);
                stack.push(sign);
                res = 0;
                sign = 1;
            } else if (c == ')') {
                res += num * sign;
                num = 0;
                res *= stack.pop(); // pop previous sign, for example: +/-(res)
                res += stack.pop(); // pop previous res, previous num + (res)
            }
        }

        res += num * sign;
        return res;
    }
}
// @lc code=end

File: 224.basic-calculator.cpp
/*
 * @lc app=leetcode id=224 lang=cpp
 *
 * [224] Basic Calculator
 */

// @lc code=start
class Solution {
public:
    int calculate(string s) {
        return calculate(s, 0).first;
    }

    pair<int, int> calculate(string s, int start) {
        stack<int> st;

        pair<int, int> res = {0, 0};
        
        char sign = '+';
        int num = 0;
        int preNum = 0;

        for (int i = start; i < s.size(); i++) {
            char c = s[i];

            if (isDigit(c)) {
                num = num * 10 + (c - '0');
            }

            if ((!isDigit(c) && c != ' ') || i == s.size() - 1) {
                if (c == '(') {
                    pair<int, int> subRes = calculate(s, i + 1);
                    num = subRes.first;

                    // store a char that is end of a ')'
                    // and then we let it keep going do following 
                    // condition, the reason is that it in this round 
                    // it will not enter the sign condition because
                    // current c == '(', so it will pass in the round
                    i = subRes.second; 
                } 
               
                switch (sign) {
                    case '+':
                        st.push(num);
                        break;
                    case '-':
                        st.push(-num);
                        break;
                    case '*':
                        preNum = st.top(); st.pop();
                        st.push(preNum * num);
                        break;
                    case '/':
                        preNum = st.top(); st.pop();
                        st.push(preNum / num);
                        break;
                }

                if (c == ')') {
                    res.second = i;
                    break;
                }

                sign = c;
                num = 0;
            }
        }

        int sum = 0;
        while (!st.empty()) {
            sum += st.top(); st.pop();
        }
        res.first = sum;

        return res;
    }

    bool isDigit(char c) {
        return '0' <= c && c <= '9';
    }
};
// @lc code=end

